File: python/ssp.py
import os
import sys
import random

class SSP:
    MAX_SOLUTIONS = 10000000

    def __init__(self, arg):
        self.solutions = []
        if isinstance(arg, int):
            n = arg
            if n <= 0:
                raise ValueError("SSP size cannot be nonpositive")
            if n <= 2:
                raise ValueError("SSP size is too small")
            self.target = 1
            self.original = [0] * n
            self.original[0] = 1
            for i in range(1, n):
                self.original[i] = i + 1
                if random.choice([True, False]):
                    self.target += self.original[i]
            # Tri décroissant pour améliorer l'élagage
            self.original.sort(reverse=True)
        elif isinstance(arg, str):
            filename = arg
            if filename is None:
                raise ValueError("Given path/file is null")
            if not os.path.exists(filename):
                raise ValueError("Given path/file does not exist")
            with open(filename, 'r') as f:
                lines = f.read().split()
            if len(lines) < 2:
                raise ValueError("Error while parsing input file")
            size = int(lines[0])
            if size <= 0:
                raise ValueError("Error while parsing input file")
            self.original = [0] * size
            self.target = int(lines[1])
            if len(lines) < 2 + size:
                raise ValueError("Error while parsing input file")
            for i in range(size):
                self.original[i] = int(lines[2 + i])
            # Tri décroissant pour améliorer l'élagage
            self.original.sort(reverse=True)
        else:
            raise TypeError("Invalid argument type for SSP constructor")

    def total_sum(self):
        return sum(self.original)

    def bp(self, i=0, partial=0, total=None, x=None):
        if total is None:
            total = self.total_sum()
        if x is None:
            x = [False] * len(self.original)
        if partial + total < self.target or partial > self.target:
            return
        if partial == self.target:
            if len(self.solutions) >= self.MAX_SOLUTIONS:
                return
            sol = [self.original[k] for k in range(i) if x[k]]
            self.solutions.append(sol)
            return
        if i == len(self.original):
            return
        total -= self.original[i]
        x[i] = False
        self.bp(i + 1, partial, total, x)
        x[i] = True
        partial += self.original[i]
        self.bp(i + 1, partial, total, x)
        x[i] = False

    def __str__(self):
        return f"SSP(n = {len(self.original)}; target = {self.target})"

    def show_target(self):
        return f"Target is {self.target}"

    def show_integers(self):
        s = "Original set = ["
        s += ",".join(str(x) for x in self.original)
        s += "]"
        return s

def print_usage():
    print("Usage: python SSP.py <integer|filename|directory>")

def is_integer(s):
    try:
        int(s)
        return True
    except ValueError:
        return False

def process_ssp(ssp):
    print(ssp)
    print(ssp.show_integers())
    print(ssp.show_target())
    print()
    print("Running bp ... ", end="")
    import time
    start = time.time()
    ssp.bp()
    end = time.time()
    print("done!")
    if len(ssp.solutions) < 10:
        print(ssp.solutions)
    else:
        print(f"bp found {len(ssp.solutions)} solutions")
    print(f"elapsed time {int((end - start) * 1000)}")
    print()

def process_directory(dir_path):
    if not os.path.isdir(dir_path):
        print(f"Le répertoire est vide ou inaccessible: {dir_path}", file=sys.stderr)
        return
    for filename in os.listdir(dir_path):
        if filename.endswith(".txt"):
            print(f"Traitement du fichier : {filename}")
            try:
                ssp = SSP(os.path.join(dir_path, filename))
                process_ssp(ssp)
            except Exception as e:
                print(f"Erreur lors du traitement du fichier {filename} : {e}", file=sys.stderr)

if __name__ == "__main__":
    if len(sys.argv) == 1:
        print_usage()
        sys.exit(0)
    arg = sys.argv[1]
    if is_integer(arg):
        ssp = SSP(int(arg))
        process_ssp(ssp)
    elif os.path.exists(arg):
        if os.path.isfile(arg):
            ssp = SSP(arg)
            process_ssp(ssp)
        elif os.path.isdir(arg):
            process_directory(arg)
        else:
            print(f"Argument is not a file or directory: {arg}", file=sys.stderr)
    else:
        print(f"Argument is not an integer or a valid file/directory: {arg}", file=sys.stderr)

File: java/SSP.java

/* Subset Sum in Java
 *
 * MJ
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.Random;
import java.util.Scanner;
import java.util.List;
import java.util.ArrayList;
import java.util.LinkedList;

public class SSP
{
   // attributs
   private long target;
   private long[] original;
   private List<ArrayList<Long>> solutions;

   // constructor (easy instances)
   public SSP(int n) throws IllegalArgumentException
   {
      if (n <= 0) throw new IllegalArgumentException("SSP size cannot be nonpositive");
      if (n <= 2) throw new IllegalArgumentException("SSP size is too small");
      Random R = new Random();
      this.target = 1;
      this.original = new long[n];
      this.original[0] = 1;
      for (int i = 1; i < n; i++)
      {
         this.original[i] = i + 1;
         if (R.nextBoolean())  this.target = this.target + this.original[i];
      }
      // Tri décroissant pour améliorer la prune
      java.util.Arrays.sort(this.original);
      for (int i = 0; i < n / 2; i++) {
         long temp = this.original[i];
         this.original[i] = this.original[n - 1 - i];
         this.original[n - 1 - i] = temp;
      }
      solutions = new LinkedList<ArrayList<Long>> ();
   }

   // constructor (from text file)
   public SSP(String filename) throws IllegalArgumentException, FileNotFoundException
   {
      if (filename == null) throw new IllegalArgumentException("Given path/file is null");
      File input = new File(filename);
      if (!input.exists()) throw new IllegalArgumentException("Given path/file does not exist");
      Scanner scan = new Scanner(input);
      IllegalArgumentException IAE = new IllegalArgumentException("Error while parsing input file");

      // number of elements in the original set
      if (!scan.hasNext()) throw IAE;
      int size = scan.nextInt();
      if (size <= 0) throw IAE;
      this.original = new long[size];

      // target
      if (!scan.hasNext()) throw IAE;
      this.target = scan.nextLong();

      // the original set
      for (int i = 0; i < size; i++)
      {
         if (!scan.hasNext()) throw IAE;
         this.original[i] = scan.nextLong();
      }
      // Tri décroissant pour améliorer la prune
      java.util.Arrays.sort(this.original);
      int n = this.original.length;
      for (int i = 0; i < n / 2; i++) {
         long temp = this.original[i];
         this.original[i] = this.original[n - 1 - i];
         this.original[n - 1 - i] = temp;
      }
      scan.close();
      solutions = new LinkedList<ArrayList<Long>> ();
   }

   // computing the sum of all integers
   public long totalSum()
   {
      int n = this.original.length;
      long sum = this.original[0];
      for (int i = 1; i < n; i++)  sum = sum + this.original[i];
      return sum;
   }

   // bp algorithm recursive calls (private method)
   private static final int MAX_SOLUTIONS = 10000000;

   private void bp(int i,long partial,long total,boolean[] x)
   {
      // are we out of bounds yet?
      if (partial + total < this.target || partial > this.target)  return;

      // did we find a new solution already?
      if (partial == this.target)
      {
         if (this.solutions.size() >= MAX_SOLUTIONS) return;
         // we store the solution (ie the selected integers) in the list
         ArrayList<Long> sol = new ArrayList<Long> (this.original.length);
         for (int k = 0; k < i; k++)  if (x[k])  sol.add(this.original[k]);
         this.solutions.add(sol);
         return;
      }

      // did we use all integers already?
      if (i == this.original.length)  return;
      total = total - this.original[i];

      // recursive call without original[i]
      x[i] = false;
      this.bp(i + 1,partial,total,x);

      // recursive call with original[i]
      x[i] = true;
      partial = partial + this.original[i];
      this.bp(i + 1,partial,total,x);
      x[i] = false;
   }

   // branch-and-prune (bp) algorithm
   public void bp()
   {
      int n = this.original.length;
      boolean[] x = new boolean[n];
      for (int i = 0; i < n; i++)  x[i] = false;
      long total = this.totalSum();
      this.bp(0,0L,total,x);
   }

   @Override
   public String toString()
   {
      return "SSP(n = " + this.original.length + "; target = " + this.target + ")";
   }

   // showing the target
   public String showTarget()
   {
      return "Target is " + this.target;
   }

   // showing the integers in the original set
   public String showIntegers()
   {
      String s = "Original set = [";
      for (int i = 0; i < this.original.length; i++)
      {
         s = s + this.original[i];
         if (i + 1 < this.original.length)  s = s + ",";
      }
      return s + "]";
   }

   // main

   public static void main(String[] args) throws Exception
   {
      if (args.length == 0) {
         printUsage();
         return;
      }
      String arg = args[0];
      java.io.File f = new java.io.File(arg);
      if (isInteger(arg)) {
         SSP ssp = new SSP(Integer.parseInt(arg));
         processSSP(ssp);
      } else if (f.exists()) {
         if (f.isFile()) {
            SSP ssp = new SSP(arg);
            processSSP(ssp);
         } else if (f.isDirectory()) {
            processDirectory(f);
         } else {
            System.err.println("Argument is not a file or directory: " + arg);
         }
      } else {
         System.err.println("Argument is not an integer or a valid file/directory: " + arg);
      }
   }

   private static void printUsage() {
      System.out.println("Usage: java SSP <integer|filename|directory>");
   }

   private static boolean isInteger(String s) {
      try {
         Integer.parseInt(s);
         return true;
      } catch (NumberFormatException e) {
         return false;
      }
   }

   private static void processSSP(SSP ssp) {
      System.out.println(ssp);
      System.out.println(ssp.showIntegers());
      System.out.println(ssp.showTarget());
      System.out.println();

      System.out.print("Running bp ... ");
      long start = System.currentTimeMillis();
      ssp.bp();
      long end = System.currentTimeMillis();
      System.out.println("done!");
      if (ssp.solutions.size() < 10)
         System.out.println(ssp.solutions);
      else
         System.out.println("bp found " + ssp.solutions.size() + " solutions");
      System.out.println("elapsed time " + (end - start));
      System.out.println();
   }

   private static void processDirectory(java.io.File dir) {
      java.io.File[] files = dir.listFiles();
      if (files == null) {
         System.err.println("Le répertoire est vide ou inaccessible: " + dir.getPath());
         return;
      }
      for (java.io.File file : files) {
         if (file.isFile() && file.getName().endsWith(".txt")) {
            System.out.println("Traitement du fichier : " + file.getName());
            try {
               SSP ssp = new SSP(file.getPath());
               processSSP(ssp);
            } catch (Exception e) {
               System.err.println("Erreur lors du traitement du fichier " + file.getName() + " : " + e.getMessage());
            }
         }
      }
   }
}
